src/utils.rs

```diff
@@ -1,33 +1,59 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
-/// Converts the value into a parallel iterator (if the multi-thread feature is enabled).
+/// Converts the value into a parallel iterator if the `multithread` feature is enabled.
 /// Uses the `rayon::iter::IntoParallelIterator` trait.
 #[cfg(all(
-    all(
-        any(target_os = "macos", target_os = "windows", target_os = "freebsd",),
-        feature = "multithread"
-    ),
+    feature = "multithread",
+    not(feature = "unknown-ci"),
     not(all(target_os = "macos", feature = "apple-sandbox")),
-    not(feature = "unknown-ci")
 ))]
+#[allow(dead_code)]
 pub(crate) fn into_iter<T>(val: T) -> T::Iter
 where
     T: rayon::iter::IntoParallelIterator,
 {
     val.into_par_iter()
 }
 
-/// Converts the value into a sequential iterator (if the multithread feature is disabled).
+/// Converts the value into a sequential iterator if the `multithread` feature is disabled.
 /// Uses the `std::iter::IntoIterator` trait.
+#[cfg(any(
+    not(feature = "multithread"),
+    feature = "unknown-ci",
+    all(target_os = "macos", feature = "apple-sandbox")
+))]
+#[allow(dead_code)]
+pub(crate) fn into_iter<T>(val: T) -> T::IntoIter
+where
+    T: IntoIterator,
+{
+    val.into_iter()
+}
+
+/// Converts the value into a parallel mutable iterator if the `multithread` feature is enabled.
+/// Uses the `rayon::iter::IntoParallelRefMutIterator` trait.
 #[cfg(all(
-    all(
-        any(target_os = "macos", target_os = "windows", target_os = "freebsd",),
-        not(feature = "multithread")
-    ),
+    feature = "multithread",
     not(feature = "unknown-ci"),
-    not(all(target_os = "macos", feature = "apple-sandbox"))
+    not(all(target_os = "macos", feature = "apple-sandbox")),
 ))]
-pub(crate) fn into_iter<T>(val: T) -> T::IntoIter
+pub(crate) fn into_iter_mut<'a, T>(
+    val: &'a mut T,
+) -> <T as rayon::iter::IntoParallelRefMutIterator<'a>>::Iter
+where
+    T: rayon::iter::IntoParallelRefMutIterator<'a> + ?Sized,
+{
+    val.par_iter_mut()
+}
+
+/// Converts the value into a sequential mutable iterator if the `multithread` feature is disabled.
+/// Uses the `std::iter::IntoIterator` trait.
+#[cfg(any(
+    not(feature = "multithread"),
+    feature = "unknown-ci",
+    all(target_os = "macos", feature = "apple-sandbox")
+))]
+pub(crate) fn into_iter_mut<T>(val: T) -> T::IntoIter
```

Comment:
Shouldn't it be:

```suggestion
pub(crate) fn into_iter_mut<T>(val: &mut T) -> T::IntoIter
```

to enforce that `val` is mutably borrowed? It would also make it more coherent with the `rayon` function.

Reply:
It is slightly confusing but in the multithreaded case the `&mut` is saying the parallel iterator is mutable.  In the non-multithreaded case the `&mut` is already part of `T` and specifying it will result in `&mut &mut T`.

Reply:
Can you add a code comment to explain that please? Like that future me will not wonder about it. :)

Reply:
Good idea.  Future me will also appreciate the clarity.